rawfile/bd2fs.py

```diff
@@ -249,19 +251,36 @@ def ControllerExpandVolume(self, request, context):
 
     @log_grpc_request
     def CreateSnapshot(self, request: csi_pb2.CreateSnapshotRequest, context):
-        volume_id = request.source_volume_id
-        name = request.name
-
-        snapshot_id, creation_time_ns = btrfs_create_snapshot(
-            volume_id=volume_id, name=name
-        )
-
+        fs = None
+        loop_dev = None
+        try:
+            file = img_file(request.source_volume_id)
+            loop_dev = attach_loop(file)
+            fs = from_device(loop_dev)
+            if not fs:
+                raise UnknownFileSystemError(
+                    device=loop_dev, volume_id=request.source_volume_id
+                )
+
+            fs.create_snapshot(name=request.name)
+            creation_time_ns = time.time_ns()
+            snapshot_id = f"{request.source_volume_id}/{request.name}"
+        finally:
+            if fs:
+                fs.unmount(clear_mountpoint=True)
+            if loop_dev:
+                subprocess.run(
```

Comment:
I think I just realized we may not be able to do this, what if the volume was actually mounted already?

Reply:
hmm I thought if device already exists it's reused for idempotency

Reply:
This is checking already existing attached devices first?

```python
            devs = attached_loops(file)
            if len(devs) > 0:
                # we could use -L to ensure there's no overlap, and thus having
                # only 1 device at most a match and allowing us to simply use
                # losetup --direct-io=on -fL --show {file}
                dev = devs[0]
                # sometimes a RO attribute is sticky on the loop device, for some reason
                run(f"blockdev --setrw {dev}")
                return dev
```